src/serac/physics/utilities/quadrature_data.hpp

```diff
@@ -200,49 +273,95 @@ class QuadratureData : public SyncableData {
+  // Doesn't do anything
+public:
+  QuadratureDataView(QuadratureData<void>& = dummy_qdata) {}
+  SERAC_HOST_DEVICE std::nullptr_t operator()(const int, const int) { return nullptr; }
+};
+
+// A dummy global so that lvalue references can be bound to something of type QData<void>
+// FIXME: There's probably a cleaner way to do this, it's technically a non-const global
+// but it's not really mutable because no operations are defined for it
+extern QuadratureDataView<void> dummy_qdata_view;
+
 // Hijacks the "vdim" parameter (number of doubles per qpt) to allocate the correct amount of storage
 template <typename T>
 QuadratureData<T>::QuadratureData(mfem::Mesh& mesh, const int p, const bool alloc)
     : qspace_(std::make_unique<mfem::QuadratureSpace>(&mesh, p + 1)),
       // When left unallocated, the allocation can happen inside the datastore
       // Use a raw pointer here when unallocated, lifetime will be managed by the DataCollection
       qfunc_(detail::initialQuadFunc(&detail::retrieve(qspace_), alloc, stride_)),
-      data_(static_cast<std::size_t>(detail::retrieve(qfunc_).Size() / stride_))
+      data_(static_cast<std::size_t>(detail::retrieve(qfunc_).Size() / stride_)),
+      offsets_(static_cast<std::size_t>(mesh.GetNE() + 1))
 {
   // To avoid violating C++'s strict aliasing rule we need to std::memcpy a default-constructed object
   // See e.g. https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8
   // also https://en.cppreference.com/w/cpp/numeric/bit_cast
   // also https://chromium.googlesource.com/chromium/src/base/+/refs/heads/master/bit_cast.h
   static_assert(std::is_default_constructible_v<T>, "Must be able to default-construct the stored type");
   static_assert(std::is_trivially_copyable_v<T>, "Uses memcpy - requires trivial copies");
+  // FIXME: Can we avoid storing a copy of the offsets array in the general case?
+  std::memcpy(offsets_.data(), detail::quadSpaceOffsets(detail::retrieve(qspace_)),
+              static_cast<std::size_t>(mesh.GetNE() + 1) * sizeof(int));
+#ifdef SERAC_USE_CHAI
+  // Unlike std::vector, ManagedArray does not appear to default-construct the elements
+  std::fill_n(data_.data(), data_.size(), T{});
+#endif
 }
 
 template <typename T>
-T& QuadratureData<T>::operator()(const int element_idx, const int q_idx)
+QuadratureData<T>& QuadratureData<T>::operator=(const T& item)
 {
-  // A view into the quadrature point data
-  mfem::Vector view;
-  // Use the existing MFEM offset calculation logic instead of reimplementing it here
-  // Avoids making this code dependent on mfem::QuadratureSpace impl
-  detail::retrieve(qfunc_).GetElementValues(element_idx, q_idx, view);
-  double*    end_ptr   = view.GetData();
-  double*    start_ptr = detail::retrieve(qfunc_).GetData();
-  const auto idx       = static_cast<std::size_t>(end_ptr - start_ptr) / stride_;
-  return data_[static_cast<std::size_t>(idx)];
+  for (auto& datum : data_) {
+    datum = item;
+  }
```

Comment:
The data is in managed memory so this will just result in a transfer - this function is not marked `__host__ __device__` though, but it definitely could be.  Do you anticipate it being useful to "assign" to the whole array from device code?

Reply:
Sorry, my comment was unclear. I don't think there's any reason to `__device__` qualify it, but directly dereferencing a device pointer from the host will fail unless it was allocated with `cudaMallocManaged()`. I thought I remembered from a previous meeting with some of the BLAST folks that `chai::ManagedArray` (confusingly) didn't actually call through to `cudaMallocManaged()` to allocate memory.

Is this function tested and working without errors when using a GPU?